sklearn/preprocessing/target.py

```diff
@@ -0,0 +1,225 @@
+        If None, the random number generator is the RandomState instance used
+        by `np.random`. The random state is used for subsampling in
+        ``check_inverse``.
+
+    Attributes
+    ----------
+    regressor_ : object
+        Fitted regressor.
+
+    transformer_ : object
+        Used transformer in ``fit`` and ``predict``.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.linear_model import LinearRegression
+    >>> from sklearn.preprocessing import TransformTargetRegressor
+    >>> tt = TransformTargetRegressor(regressor=LinearRegression(),
+    ...                                 func=np.log, inverse_func=np.exp)
+    >>> X = np.arange(4).reshape(-1, 1)
+    >>> y = np.exp(2 * X).ravel()
+    >>> tt.fit(X, y)
+    ... #doctest: +NORMALIZE_WHITESPACE
+    TransformTargetRegressor(check_inverse=True,
+                             func=<ufunc 'log'>,
+                             inverse_func=<ufunc 'exp'>,
+                             random_state=None,
+                             regressor=LinearRegression(copy_X=True,
+                                                        fit_intercept=True,
+                                                        n_jobs=1,
+                                                        normalize=False),
+                             transformer=None)
+    >>> tt.score(X, y)
+    1.0
+    >>> tt.regressor_.coef_
+    array([ 2.])
+
+    Notes
+    -----
+    The benefit of transforming the target before fitting a regression model is
+    highlighted in :ref:`examples/preprocessing/plot_transform_target.py
+    <sphx_glr_auto_examples_preprocessing_plot_transform_target.py> `.
+
+    See also :ref:`User guide <scaling_regression>`
+
+    """
+    def __init__(self, regressor=None, transformer=None,
+                 func=None, inverse_func=None, check_inverse=True,
+                 random_state=None):
+        self.regressor = regressor
+        self.transformer = transformer
+        self.func = func
+        self.inverse_func = inverse_func
+        self.check_inverse = check_inverse
+        self.random_state = random_state
+
+    def _fit_transformer(self, y, sample_weight):
+        if (self.transformer is not None and
+                (self.func is not None or self.inverse_func is not None)):
+            raise ValueError("Both 'transformer' and functions 'func'/"
+                             "'inverse_func' cannot be set at the same time.")
+        elif self.transformer is not None:
+            self.transformer_ = clone(self.transformer)
+        else:
+            self.transformer_ = FunctionTransformer(
+                func=self.func, inverse_func=self.inverse_func, validate=False)
+        fit_parameters = signature(self.transformer_.fit).parameters
+        if "sample_weight" in fit_parameters:
+            self.transformer_.fit(y, sample_weight=sample_weight)
+        else:
+            self.transformer_.fit(y)
+        if self.check_inverse:
```

Comment:
does it make sense to move this check inside `FunctionTransformer`? (at some point, maybe not here)